plugins/org.eclipse.glsp.server/src/org/eclipse/glsp/server/features/core/model/ModelSubmissionHandler.java

```diff
@@ -87,15 +89,44 @@
     * </p>
     *
     * @param modelState The model state to submit.
+    * @param causedBy   The optional action/operation that caused the model update.
     * @return A list of actions to be processed in order to submit the model.
     */
-   public List<Action> submitModelDirectly(final GModelState modelState) {
+   public List<Action> submitModelDirectly(final GModelState modelState, final Action causedBy) {
       GModelRoot gModel = modelState.getRoot();
-      Action action = gModel.getRevision() == 0 ? new SetModelAction(gModel) : new UpdateModelAction(gModel);
+      DiagramConfiguration diagramConfiguration = diagramConfigurationRegistry.get(modelState);
+      if (diagramConfiguration.getLayoutKind() == ServerLayoutKind.AUTOMATIC) {
+         layoutEngine.layout(modelState);
+      }
+      Action modelAction = gModel.getRevision() == 0 ? new SetModelAction(gModel) : new UpdateModelAction(gModel);
       synchronized (modelLock) {
-         return Arrays.asList(action, new SetDirtyStateAction(modelState.isDirty()));
+         List<Action> result = new ArrayList<>();
+         result.add(modelAction);
+         Optional.ofNullable(causedBy).ifPresent(result::add);
```

Comment:
I'm not sure I understand why we would need to add the 'cause' to the result? The method comment mentions that the result of this call should be dispatched through the ActionDispatcher (either directly or not), would that lead not to an infinite loop in case we do not need a client layout or do I miss something here?